sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -892,132 +892,108 @@ case class MapFromEntries(child: Expression)
-
-    val mapData = array.asInstanceOf[MapData]
-    val numElements = mapData.numElements()
-    val keys = mapData.keyArray()
-    val values = mapData.valueArray()
-
-    val ordering = PhysicalDataType.ordering(keyType)
-
-    val sortedMap = Array
-      .tabulate(numElements)(i => (keys.get(i, keyType).asInstanceOf[Any],
-        values.get(i, valueType).asInstanceOf[Any]))
-      .sortBy(_._1)(ordering)
-
-    new ArrayBasedMapData(new GenericArrayData(sortedMap.map(_._1)),
-      new GenericArrayData(sortedMap.map(_._2)))
-  }
-
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    nullSafeCodeGen(ctx, ev, b => sortCodegen(ctx, ev, b))
-  }
-
-  private def sortCodegen(ctx: CodegenContext, ev: ExprCode,
-      base: String): String = {
-
-    val arrayBasedMapData = classOf[ArrayBasedMapData].getName
-    val genericArrayData = classOf[GenericArrayData].getName
-
-    val numElements = ctx.freshName("numElements")
-    val keys = ctx.freshName("keys")
-    val values = ctx.freshName("values")
-    val sortArray = ctx.freshName("sortArray")
-    val i = ctx.freshName("i")
-    val o1 = ctx.freshName("o1")
-    val o1entry = ctx.freshName("o1entry")
-    val o2 = ctx.freshName("o2")
-    val o2entry = ctx.freshName("o2entry")
-    val c = ctx.freshName("c")
-    val newKeys = ctx.freshName("newKeys")
-    val newValues = ctx.freshName("newValues")
-
-    val boxedKeyType = CodeGenerator.boxedType(keyType)
-    val boxedValueType = CodeGenerator.boxedType(valueType)
-    val javaKeyType = CodeGenerator.javaType(keyType)
+    }
 
-    val simpleEntryType = s"java.util.AbstractMap.SimpleEntry<$boxedKeyType, $boxedValueType>"
+    if (dataType.existsRecursively(dt =>
+      dt.isInstanceOf[MapType] && !RowOrdering.isOrderable(dt.asInstanceOf[MapType].keyType))) {
+      DataTypeMismatch(
+        errorSubClass = "INVALID_ORDERING_TYPE",
+        messageParameters = Map(
+          "functionName" -> toSQLId(prettyName),
+          "dataType" -> toSQLType(dataType)
+        )
+      )
+    }
 
-    val comp = if (CodeGenerator.isPrimitiveType(keyType)) {
-      val v1 = ctx.freshName("v1")
-      val v2 = ctx.freshName("v2")
-      s"""
-         |$javaKeyType $v1 = (($boxedKeyType) $o1).${javaKeyType}Value();
-         |$javaKeyType $v2 = (($boxedKeyType) $o2).${javaKeyType}Value();
-         |int $c = ${ctx.genComp(keyType, v1, v2)};
-       """.stripMargin
-    } else {
-      s"int $c = ${ctx.genComp(keyType, s"(($javaKeyType) $o1)", s"(($javaKeyType) $o2)")};"
+    TypeCheckResult.TypeCheckSuccess
+  }
+
+  // Evaluates the expression recursively by taking into
+  // account complex types and nesting
+  def nullSafeEvalRecursive(input: Any, dataType: DataType): Any = {
+
+    dataType match {
+      // For ArrayType recursively call evaluate for
+      // all its children since MapType can be nested
+      // as array element
+      case ArrayType(elementType, _) =>
```

Comment:
We should reuse existing expressions to do the recursion over complicated types. See `CharVarcharUtils.stringLengthCheck` as an example, we use `ArrayTransform` expression to handle elements of array.

Reply:
We are working here with input data, but ArrayTransform is expecting expression as an input, is this doable?

Reply:
It's not doable here. We should do it in the rule where we add `MapSort` expression.

Reply:
What about the case where i.e. ArrayType is nested into StructType? 
Something like StructType<a:ArrayType<MapType<INT,INT>>>

Reply:
We know the type information at query plan compile time, we can always build the correct expression to handle the actual nested types. `CharVarcharUtils.stringLengthCheck` handles struct type, isn't it?